rate_limit_quotas.py

```diff
@@ -0,0 +1,186 @@
+"""
+Rate limiter and quota system.
+
+Framework-independent rate limiting mechanism that provides:
+    * IP address and token-based accounting
+    * customizable quotas based on IP address and token
+    * late limiting based on resource usage (time spent on API calls)
+    * bucketing based on day, week, month
+    * statistics
+    * metrics
+    * fast in-memory storage
+
+Also provides a connector for Flask
+
+"""
+
+import time
+import ipaddress
+from typing import Dict
+
+
+class Limiter:
+    def __init__(
+        self,
+        limits: dict,
+        token_check_callback=None,
+        ipaddr_methods=["X-Real-Ip", "socket"],
+    ):
+        # Bucket sequence: month, week, day
+        self._ipaddr_limits = [
+            limits.get(l, None)
+            for l in ("ipaddr_per_month", "ipaddr_per_week", "ipaddr_per_day")
+        ]
+        self._token_limits = [
+            limits.get(l, None)
+            for l in ("token_per_month", "token_per_week", "token_per_day")
+        ]
+        self._ipaddr_buckets = ({}, {}, {})  # type: Dict
+        self._token_buckets = ({}, {}, {})  # type: Dict
+        self._token_check_callback = token_check_callback
+        self._ipaddr_extraction_methods = ipaddr_methods
+        self._last_quota_update_time = time.monotonic()
+        self.increment_quota_counters(1)
+        self.refresh_quota_counters_if_needed()
+
+    def increment_quota_counters(self, tdelta: int):
+        """Delta: time from previous run in seconds"""
+        if tdelta <= 0:
+            return
+
+        iterable = (
+            (30 * 24, self._ipaddr_limits[0], self._ipaddr_buckets[0]),
+            (7 * 24, self._ipaddr_limits[0], self._ipaddr_buckets[0]),
+            (1 * 24, self._ipaddr_limits[0], self._ipaddr_buckets[0]),
+            (30 * 24, self._ipaddr_limits[0], self._ipaddr_buckets[0]),
+            (7 * 24, self._ipaddr_limits[0], self._ipaddr_buckets[0]),
+            (1 * 24, self._ipaddr_limits[0], self._ipaddr_buckets[0]),
+        )
+        for hours, limit, bucket in iterable:
+            vdelta = limit / hours / 3600 * tdelta
+            to_delete = []
+            for k, v in bucket.items():
+                v += vdelta
+                if v >= limit:
+                    to_delete.append(k)
+                else:
+                    bucket[k] = v
+
+            for k in to_delete:
+                del bucket[k]
+
+    def refresh_quota_counters_if_needed(self):
```

Comment:
I see that this is only called in the constructor, shouldn't it be called also periodically in some way?

Reply:
Yes, this will be added externally to the module.